### PR TITLE
feat: testscript POC (DO NOT MERGE)

### PR DESCRIPTION
**Pull request by ashearin**
## Description

POC of testscript IT, for illustration/comparison purposes

### CODE REVIEW

File: internal/e2e/e2eutil/check_files.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"path"
+
+	"github.com/rogpeppe/go-internal/testscript"
+)
+
+// checkFile is a testscript command that checks the
+// existence of a list of files inside a directory.
+func checkFiles(script *testscript.TestScript, neg bool, args []string) { //nolint: revive
```
Comment (jhoward-lm) [inline internal/e2e/e2eutil/check_files.go:30]: Does it have to match this signature? Or would this work. I also think I remember reading somewhere that `//nolint` directives shouldn't have a space after the `:`

```suggestion
func checkFiles(script *testscript.TestScript, neg bool, args ...string) { //nolint:revive
```

If it has to match this signature, and it's used in multiple places, can we define a type with it?

```go
type commandFunc func(*testscript.TestScript, bool, []string)
```

File: internal/e2e/e2eutil/custom_commands.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// -----------------------------------------------------------------------------
+
+package e2eutil
+
+import "github.com/rogpeppe/go-internal/testscript"
+
+func CustomCommands() map[string]func(ts *testscript.TestScript, neg bool, args []string) {
+	return map[string]func(ts *testscript.TestScript, neg bool, args []string){
```
Comment (jhoward-lm) [inline internal/e2e/e2eutil/custom_commands.go:25]: Same question as above about the signature

Comment (jhoward-lm) [inline internal/e2e/e2eutil/custom_commands.go:25]: Shouldn't need to name the params if just used as a return type

```suggestion
func CustomCommands() map[string]func(*testscript.TestScript, bool, []string) {
	return map[string]func(*testscript.TestScript, bool, []string){
```

File: internal/e2e/e2eutil/custom_conditions.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		elapsed++
+	}
+
+	return false, errFileNotFound
+}
+
+func fileExists(filename string) bool {
+	_, err := os.Stat(filename)
+
+	return !os.IsNotExist(err)
```
Comment (jhoward-lm) [inline internal/e2e/e2eutil/custom_conditions.go:129]: From the [`os` docs](https://pkg.go.dev/os#IsNotExist)

```suggestion
	return !errors.Is(err, fs.ErrNotExist)
```

File: cmd/root.go [hunk 4]
```diff
@@ -172,6 +172,16 @@ func rootCmd() *cobra.Command {
...
+// }
+
+// control your exit status with Execute, by default in cmd/root.go:.
+func Execute() int {
+	if err := rootCmd().Execute(); err != nil {
+		// log it, then
+		return 1
+	}
+
+	return 0
```
Comment (jhoward-lm) [inline cmd/root.go:186]: Doesn't main.go need updating?

There are several possible panics or `log.Fatal` calls in the subcommands. Should this also have a recover handler?

```suggestion
func Execute() (rc int) {
	defer func() {
		if r := recover(); r != nil {
			// handle the panic
			rc = 1
		}
	}()

	if err := rootCmd().Execute(); err != nil {
		rc = 1
	}

	return
```

File: internal/e2e/e2eutil/check_files.go (hunk 1 above)
Comment (ashearin) [reply internal/e2e/e2eutil/check_files.go:30]: you are right about the no lint thing, however we have two rules that argue over that, so I've been needing to put two nolint entries to get it to ignore the line

**Pull request closed without merging**